File: apps/mobile/lib/screens/chat_screen.dart
// chat_screen.dart — conversación con streaming SSE, markdown en respuestas
// y selector de modelo. Burbujas: usuario invertido (primary), asistente
// plano, como en el chat de la web.
import 'package:flutter/material.dart';
import 'package:flutter_markdown/flutter_markdown.dart';
import 'package:provider/provider.dart';
import 'package:url_launcher/url_launcher.dart';

import '../state.dart';
import '../theme.dart';
import '../widgets/common.dart';

class ChatScreen extends StatefulWidget {
  const ChatScreen({super.key});

  @override
  State<ChatScreen> createState() => _ChatScreenState();
}

class _ChatScreenState extends State<ChatScreen> {
  late final TextEditingController input;

  @override
  void initState() {
    super.initState();
    final chat = context.read<ChatState>();
    input = TextEditingController(text: chat.draft);
    if (chat.models.isEmpty) chat.loadModels();
  }

  @override
  void dispose() {
    input.dispose();
    super.dispose();
  }

  void _send() {
    final chat = context.read<ChatState>();
    final text = input.text.trim();
    if (text.isEmpty || chat.streaming) return;
    input.clear();
    chat.send(text);
  }

  Future<void> _pickModel() async {
    final chat = context.read<ChatState>();
    final c = LixColors.of(context);
    await showModalBottomSheet<void>(
      context: context,
      backgroundColor: c.bg,
      shape: const RoundedRectangleBorder(
        borderRadius: BorderRadius.vertical(top: Radius.circular(kRadiusBox)),
      ),
      builder: (ctx) => SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(18),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                'Modelo',
                style: TextStyle(
                  color: c.inkSoft,
                  fontFamily: kFontUi,
                  fontWeight: FontWeight.w600,
                  fontSize: 13,
                ),
              ),
              const SizedBox(height: 8),
              if (chat.models.isEmpty)
                Text(
                  'No hay modelos disponibles.',
                  style: TextStyle(color: c.inkMuted, fontFamily: kFontUi, fontSize: 13),
                ),
              ...chat.models.map((m) => ListTile(
                    dense: true,
                    shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
                    tileColor: m == chat.model ? c.accentSoft : null,
                    title: Text(
                      m,
                      style: TextStyle(
                        color: c.ink,
                        fontFamily: kFontUi,
                        fontWeight: m == chat.model ? FontWeight.w500 : FontWeight.w400,
                        fontSize: 14,
                      ),
                    ),
                    trailing: m == chat.model
                        ? Icon(Icons.check, size: 16, color: c.accentDeep)
                        : null,
                    onTap: () {
                      chat.setModel(m);
                      Navigator.of(ctx).pop();
                    },
                  )),
            ],
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final c = LixColors.of(context);
    final chat = context.watch<ChatState>();
    final auth = context.watch<AuthState>();
    final firstName = auth.user?['first_name'];

    return Scaffold(
      backgroundColor: c.bg,
      body: SafeArea(
        child: Column(
          children: [
            // Cabecera: título + modelo + nueva conversación
            Container(
              padding: const EdgeInsets.fromLTRB(16, 8, 12, 10),
              decoration: BoxDecoration(
                border: Border(bottom: BorderSide(color: c.borderSoft)),
              ),
              child: Row(
                children: [
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          chat.title ?? 'Nueva conversación',
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          style: TextStyle(
                            color: c.ink,
                            fontFamily: kFontUi,
                            fontWeight: FontWeight.w600,
                            fontSize: 16,
                          ),
                        ),
                        InkWell(
                          onTap: _pickModel,
                          child: Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              Flexible(
                                child: Text(
                                  chat.model.isEmpty ? 'Sin modelos' : chat.model,
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                  style: TextStyle(
                                    color: c.inkSoft,
                                    fontFamily: kFontUi,
                                    fontSize: 12,
                                  ),
                                ),
                              ),
                              Icon(Icons.keyboard_arrow_down, size: 14, color: c.inkSoft),
                            ],
                          ),
                        ),
                      ],
                    ),
                  ),
                  IconButton(
                    onPressed: chat.newChat,
                    style: IconButton.styleFrom(backgroundColor: c.bgSecondary),
                    icon: Icon(Icons.edit_outlined, size: 20, color: c.ink),
                    tooltip: 'Nueva conversación',
                  ),
                ],
              ),
            ),

            Expanded(
              child: chat.loadingMessages
                  ? Center(child: CircularProgressIndicator(color: c.inkSoft))
                  : chat.messages.isEmpty
                      ? _EmptyState(firstName: firstName is String ? firstName : null)
                      : _MessageList(chat: chat),
            ),

            if (chat.error.isNotEmpty)
              Container(
                margin: const EdgeInsets.fromLTRB(16, 0, 16, 6),
                padding: const EdgeInsets.all(10),
                decoration: BoxDecoration(
                  color: c.dangerSoft,
                  borderRadius: BorderRadius.circular(12),
                ),
                child: Row(
                  children: [
                    Expanded(
                      child: Text(
                        chat.error,
                        style: TextStyle(color: c.danger, fontFamily: kFontUi, fontSize: 13),
                      ),
                    ),
                    InkWell(
                      onTap: chat.clearError,
                      child: Icon(Icons.close, size: 16, color: c.danger),
                    ),
                  ],
                ),
              ),

            // Barra de composición
            Padding(
              padding: const EdgeInsets.fromLTRB(12, 6, 12, 10),
              child: Container(
                padding: const EdgeInsets.all(6),
                decoration: BoxDecoration(
                  color: c.bgInput,
                  borderRadius: BorderRadius.circular(26),
                  border: Border.all(color: c.borderSoft),
                ),
                child: Row(
                  crossAxisAlignment: CrossAxisAlignment.end,
                  children: [
                    // Modo investigar (búsqueda web)
                    InkWell(
                      borderRadius: BorderRadius.circular(18),
                      onTap: () => chat.setWebSearch(!chat.webSearch),
                      child: Container(
                        width: 36,
                        height: 36,
                        decoration: BoxDecoration(
                          color: chat.webSearch ? c.accentSoft : Colors.transparent,
                          shape: BoxShape.circle,
                        ),
                        child: Icon(
                          Icons.public,
                          size: 19,
                          color: chat.webSearch ? c.accentDeep : c.inkMuted,
                        ),
                      ),
                    ),
                    const SizedBox(width: 6),
                    Expanded(
                      child: TextField(
                        controller: input,
                        onChanged: chat.setDraft,
                        minLines: 1,
                        maxLines: 5,
                        textInputAction: TextInputAction.newline,
                        style: TextStyle(color: c.ink, fontFamily: kFontUi, fontSize: 15),
                        decoration: InputDecoration(
                          hintText: 'Escribe un mensaje…',
                          hintStyle: TextStyle(color: c.inkMuted, fontFamily: kFontUi),
                          border: InputBorder.none,
                          isDense: true,
                          contentPadding: const EdgeInsets.symmetric(vertical: 9),
                        ),
                      ),
                    ),
                    const SizedBox(width: 6),
                    InkWell(
                      borderRadius: BorderRadius.circular(18),
                      onTap: chat.streaming ? chat.stop : _send,
                      child: Container(
                        width: 36,
                        height: 36,
                        decoration: BoxDecoration(color: c.primary, shape: BoxShape.circle),
                        child: Icon(
                          chat.streaming ? Icons.stop : Icons.arrow_upward,
                          size: 18,
                          color: c.onPrimary,
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _EmptyState extends StatelessWidget {
  final String? firstName;
  const _EmptyState({this.firstName});

  @override
  Widget build(BuildContext context) {
    final c = LixColors.of(context);
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Spark(size: 26, color: c.accentDeep),
          const SizedBox(height: 14),
          const LixLogo(size: 24),
          const SizedBox(height: 14),
          Text(
            '${firstName != null ? 'Hola, $firstName. ' : ''}¿En qué te ayudo hoy?',
            textAlign: TextAlign.center,
            style: TextStyle(color: c.inkSoft, fontFamily: kFontUi, fontSize: 15),
          ),
        ],
      ),
    );
  }
}

class _MessageList extends StatelessWidget {
  final ChatState chat;
  const _MessageList({required this.chat});

  @override
  Widget build(BuildContext context) {
    final c = LixColors.of(context);
    // Lista invertida: el índice 0 es el último mensaje → auto-scroll natural.
    final items = chat.messages.reversed.toList();
    return ListView.builder(
      reverse: true,
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 14),
      itemCount: items.length,
      itemBuilder: (context, index) {
        final msg = items[index];
        final isLast = index == 0;
        if (msg.role == 'user') {
          return Align(
            alignment: Alignment.centerRight,
            child: Container(
              margin: const EdgeInsets.symmetric(vertical: 6),
              padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 10),
              constraints: BoxConstraints(
                maxWidth: MediaQuery.of(context).size.width * 0.85,
              ),
              decoration: BoxDecoration(
                color: c.primary,
                borderRadius: const BorderRadius.only(
                  topLeft: Radius.circular(18),
                  topRight: Radius.circular(18),
                  bottomLeft: Radius.circular(18),
                  bottomRight: Radius.circular(6),
                ),
              ),
              child: Text(
                msg.content,
                style: TextStyle(
                  color: c.onPrimary,
                  fontFamily: kFontUi,
                  fontSize: 15,
                  height: 1.4,
                ),
              ),
            ),
          );
        }
        return Padding(
          padding: const EdgeInsets.symmetric(vertical: 6),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              if (msg.sources != null && msg.sources!.isNotEmpty)
                Padding(
                  padding: const EdgeInsets.only(bottom: 6),
                  child: Wrap(
                    spacing: 6,
                    runSpacing: 6,
                    children: msg.sources!.take(4).map((s) {
                      final title = (s is Map ? (s['title'] ?? s['url']) : null)?.toString() ?? '';
                      final url = (s is Map ? s['url'] : null)?.toString();
                      return InkWell(
                        onTap: url != null
                            ? () => launchUrl(Uri.parse(url), mode: LaunchMode.externalApplication)
                            : null,
                        child: Container(
                          padding: const EdgeInsets.symmetric(horizontal: 9, vertical: 4),
                          decoration: BoxDecoration(
                            color: c.bgSecondary,
                            borderRadius: BorderRadius.circular(kRadiusPill),
                            border: Border.all(color: c.borderSoft),
                          ),
                          child: Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              Icon(Icons.public, size: 12, color: c.inkSoft),
                              const SizedBox(width: 5),
                              ConstrainedBox(
                                constraints: const BoxConstraints(maxWidth: 140),
                                child: Text(
                                  title,
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                  style: TextStyle(
                                    color: c.inkSoft,
                                    fontFamily: kFontUi,
                                    fontSize: 11,
                                  ),
                                ),
                              ),
                            ],
                          ),
                        ),
                      );
                    }).toList(),
                  ),
                ),
              if (msg.content.isNotEmpty)
                MarkdownBody(
                  data: msg.content,
                  onTapLink: (text, href, _) {
                    if (href != null) {
                      launchUrl(Uri.parse(href), mode: LaunchMode.externalApplication);
                    }
                  },
                  styleSheet: _markdownStyle(c),
                )
              else if (isLast && chat.streaming)
                Row(
                  children: [
                    SizedBox(
                      width: 14,
                      height: 14,
                      child: CircularProgressIndicator(strokeWidth: 2, color: c.inkSoft),
                    ),
                    const SizedBox(width: 8),
                    Text(
                      'Pensando…',
                      style: TextStyle(color: c.inkMuted, fontFamily: kFontUi, fontSize: 13),
                    ),
                  ],
                ),
            ],
          ),
        );
      },
    );
  }
}

MarkdownStyleSheet _markdownStyle(LixColors c) {
  const mono = TextStyle(fontFamily: 'monospace', fontSize: 13);
  return MarkdownStyleSheet(
    p: TextStyle(color: c.ink, fontFamily: kFontUi, fontSize: 15, height: 1.45),
    h1: TextStyle(color: c.ink, fontFamily: kFontUi, fontWeight: FontWeight.w600, fontSize: 20),
    h2: TextStyle(color: c.ink, fontFamily: kFontUi, fontWeight: FontWeight.w600, fontSize: 18),
    h3: TextStyle(color: c.ink, fontFamily: kFontUi, fontWeight: FontWeight.w600, fontSize: 16),
    strong: TextStyle(color: c.ink, fontFamily: kFontUi, fontWeight: FontWeight.w700),
    em: TextStyle(color: c.ink, fontFamily: kFontUi, fontStyle: FontStyle.italic),
    a: TextStyle(color: c.accentDeep, decoration: TextDecoration.underline),
    listBullet: TextStyle(color: c.inkSoft, fontFamily: kFontUi, fontSize: 15),
    blockquote: TextStyle(color: c.inkSoft, fontFamily: kFontUi, fontSize: 14),
    blockquoteDecoration: BoxDecoration(
      color: c.bgSecondary,
      borderRadius: BorderRadius.circular(6),
      border: Border(left: BorderSide(color: c.accentDeep, width: 3)),
    ),
    // Un único estilo `code` cubre inline y bloques: fondo bgSecondary en
    // ambos para que el texto tinta sea legible en los dos temas.
    code: mono.copyWith(color: c.ink, backgroundColor: c.bgSecondary),
    codeblockDecoration: BoxDecoration(
      color: c.bgSecondary,
      borderRadius: BorderRadius.circular(12),
      border: Border.all(color: c.borderSoft),
    ),
    codeblockPadding: const EdgeInsets.all(12),
    horizontalRuleDecoration: BoxDecoration(
      border: Border(top: BorderSide(color: c.borderSoft)),
    ),
    tableBorder: TableBorder.all(color: c.borderSoft),
    tableBody: TextStyle(color: c.ink, fontFamily: kFontUi, fontSize: 13),
  );
}

File: apps/mobile/src/screens/ChatScreen.js
// ChatScreen.js — conversación con streaming SSE, markdown en respuestas y
// selector de modelo. Espejo del chat de la web (chat.css): cabecera limpia
// con el título en pill y punto de estado, hero de marca centrado
// ("¿Qué investigaremos hoy?" + atajos de arranque) cuando no hay mensajes,
// burbuja del usuario invertida (radius-box con esquina 15), y compositor caja
// (.chat-input): texto arriba y barra inferior con búsqueda web + modelo a la
// izquierda y enviar (círculo tinta 40px) a la derecha.
import React, { useState } from 'react';
import {
  ActivityIndicator,
  FlatList,
  KeyboardAvoidingView,
  Linking,
  Pressable,
  Text,
  TextInput,
  View,
} from 'react-native';
import Markdown from 'react-native-markdown-display';
import { SafeAreaView, useSafeAreaInsets } from 'react-native-safe-area-context';

import Icon from '../components/Icon';
import { useDialogs } from '../components/dialogs';
import { ChatHeader, FadeUp, useColors } from '../components/ui';
import { useApi, useAuth, useChat } from '../state';
import { FONTS, RADIUS_BOX, RADIUS_PILL } from '../theme';

// Atajos del hero: rellenan el compositor con un arranque de conversación en
// vez de enviarlo, para que el usuario complete la idea antes de disparar.
const STARTERS = [
  { icon: 'target', label: 'Investigar un tema a fondo', prompt: 'Investiga a fondo ' },
  { icon: 'doc', label: 'Analizar un documento', prompt: 'Analiza este documento y resume lo importante:\n\n' },
  { icon: 'waves', label: 'Comparar datos y fuentes', prompt: 'Compara estos datos y contrasta las fuentes:\n\n' },
];

export default function ChatScreen({ onMenu }) {
  const c = useColors();
  const chat = useChat();
  const auth = useAuth();
  const api = useApi();
  const { sheet, prompt, confirm, toast } = useDialogs();
  const [input, setInput] = useState(chat.draftRef.current);

  React.useEffect(() => {
    if (chat.models.length === 0) chat.loadModels();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Menú ⋮ de la cabecera. Las acciones sobre la conversación solo tienen
  // sentido cuando ya existe en el servidor, así que se ocultan si no la hay.
  const openOptions = async () => {
    const saved = !!chat.conversationId && chat.messages.length > 0;
    const action = await sheet({
      title: chat.title || 'Nueva conversación',
      items: [
        { label: 'Nueva conversación', icon: 'plus', value: 'new' },
        { label: 'Cambiar de modelo', icon: 'activity', value: 'model' },
        ...(saved
          ? [
              { label: 'Renombrar', icon: 'pencil', value: 'rename' },
              { label: 'Eliminar', icon: 'trash', danger: true, value: 'delete' },
            ]
          : []),
      ],
    });
    if (action === 'new') chat.newChat();
    if (action === 'model') await pickModel();
    if (action === 'rename') {
      const value = await prompt({
        title: 'Renombrar conversación',
        placeholder: 'Nuevo título',
        initialValue: chat.title || '',
        confirmLabel: 'Guardar',
      });
      const trimmed = (value || '').trim();
      if (!trimmed) return;
      try {
        await api.patch(`/api/conversations/${chat.conversationId}`, { title: trimmed });
        chat.setTitle(trimmed);
      } catch {
        toast('No se pudo renombrar la conversación');
      }
    }
    if (action === 'delete') {
      const ok = await confirm({
        title: 'Eliminar conversación',
        message: `"${chat.title || 'Sin título'}" se eliminará definitivamente.`,
        confirmLabel: 'Eliminar',
        danger: true,
      });
      if (!ok) return;
      try {
        await api.delete(`/api/conversations/${chat.conversationId}`);
        chat.newChat();
      } catch {
        toast('No se pudo eliminar la conversación');
      }
    }
  };

  const pickModel = async () => {
    const value = await sheet({
      title: 'Modelo',
      emptyLabel: 'No hay modelos disponibles.',
      items: chat.models.map((m) => ({ label: m, value: m, selected: m === chat.model })),
    });
    if (value) chat.setModel(value);
  };

  // El draft sobrevive al cambio de pestaña (vive en el contexto).
  const onChangeInput = (v) => {
    setInput(v);
    chat.draftRef.current = v;
  };

  const send = () => {
    const text = input.trim();
    if (!text || chat.streaming) return;
    setInput('');
    chat.draftRef.current = '';
    chat.send(text);
  };

  const empty = chat.messages.length === 0 && !chat.loadingMessages;
  const firstName = typeof auth.user?.first_name === 'string' ? auth.user.first_name : null;

  // Subtítulo: modelo activo, o el aviso de que está respondiendo.
  const subtitle = chat.streaming
    ? 'Respondiendo…'
    : chat.model || (chat.models.length === 0 ? 'Sin modelos' : '');

  return (
    <SafeAreaView edges={['top']} style={{ flex: 1, backgroundColor: c.bg }}>
      {/* KeyboardAvoidingView es lo que faltaba: con edge-to-edge la ventana ya
          no se redimensiona sola y el compositor quedaba DEBAJO del teclado, así
          que no se veía lo que se escribía. React Native mide el teclado con
          WindowInsets.ime() y descuenta la barra de navegación, por eso el
          inset inferior del compositor y este padding se suman sin solaparse. */}
      <KeyboardAvoidingView behavior="padding" style={{ flex: 1 }}>
        <ChatHeader
          title={chat.title || 'Nueva conversación'}
          subtitle={subtitle}
          onLeading={onMenu}
          leadingIcon="menu"
          onOptions={openOptions}
          dot={chat.streaming ? 'live' : 'idle'}
        />

        <View style={{ flex: 1 }}>
          {chat.loadingMessages ? (
            <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
              <ActivityIndicator color={c.inkSoft} />
            </View>
          ) : empty ? (
            <Hero firstName={firstName} onPick={onChangeInput} />
          ) : (
            <MessageList />
          )}
        </View>

        <Composer
          input={input}
          onChangeInput={onChangeInput}
          onSend={send}
          onPickModel={pickModel}
        />
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
}

// Hero del chat vacío (.chat-hero): marca, saludo, título grande centrado y
// atajos de arranque que precargan el compositor.
function Hero({ firstName, onPick }) {
  const c = useColors();
  return (
    <View style={{ flex: 1, justifyContent: 'center', paddingHorizontal: 26, paddingBottom: 20 }}>
      <FadeUp style={{ alignItems: 'center', gap: 8 }}>
        <View
          style={{
            width: 58,
            height: 58,
            borderRadius: 20,
            backgroundColor: c.bgSecondary,
            alignItems: 'center',
            justifyContent: 'center',
            marginBottom: 8,
          }}
        >
          <Text style={{ fontFamily: FONTS.brand, fontSize: 26, color: c.inkSoft }}>L</Text>
        </View>
        {!!firstName && (
          <Text style={{ fontFamily: FONTS.ui, fontSize: 15, color: c.inkSoft }}>
            Hola, {firstName}.
          </Text>
        )}
        <Text
          style={{
            textAlign: 'center',
            fontFamily: FONTS.uiBold,
            fontSize: 34,
            lineHeight: 40,
            letterSpacing: -0.5,
            color: c.ink,
          }}
        >
          ¿Qué investigaremos hoy?
        </Text>
      </FadeUp>

      <View style={{ gap: 10, marginTop: 34 }}>
        {STARTERS.map((s, i) => (
          <FadeUp key={s.icon} delay={80 + i * 60}>
            <Pressable
              onPress={() => onPick(s.prompt)}
              style={({ pressed }) => ({
                flexDirection: 'row',
                alignItems: 'center',
                gap: 14,
                padding: 12,
                borderRadius: 18,
                borderWidth: 1,
                borderColor: c.borderSoft,
                backgroundColor: pressed ? c.bgInput : c.bgSecondary,
              })}
            >
              <View
                style={{
                  width: 38,
                  height: 38,
                  borderRadius: 12,
                  alignItems: 'center',
                  justifyContent: 'center',
                  backgroundColor: c.bg,
                }}
              >
                <Icon name={s.icon} size={18} color={c.accent} />
              </View>
              <Text
                style={{ flex: 1, fontFamily: FONTS.uiMedium, fontSize: 14.5, lineHeight: 20, color: c.ink }}
              >
                {s.label}
              </Text>
              <Icon name="chevron-right" size={16} color={c.inkMuted} />
            </Pressable>
          </FadeUp>
        ))}
      </View>
    </View>
  );
}

function MessageList() {
  const c = useColors();
  const chat = useChat();
  // Lista invertida: el índice 0 es el último mensaje → auto-scroll natural.
  const items = [...chat.messages].reverse();

  return (
    <FlatList
      inverted
      data={items}
      keyExtractor={(_, i) => String(chat.messages.length - i)}
      contentContainerStyle={{ paddingHorizontal: 18, paddingVertical: 14 }}
      keyboardShouldPersistTaps="handled"
      renderItem={({ item, index }) => {
        const isLast = index === 0;
        if (item.role === 'user') {
          return (
            <FadeUp style={{ alignItems: 'flex-end', marginVertical: 8 }}>
              <View
                style={{
                  maxWidth: '80%',
                  backgroundColor: c.primary,
                  paddingHorizontal: 19,
                  paddingVertical: 11,
                  borderRadius: RADIUS_BOX,
                  borderBottomRightRadius: 15,
                }}
              >
                <Text style={{ fontFamily: FONTS.ui, fontSize: 15, lineHeight: 22, color: c.onPrimary }}>
                  {item.content}
                </Text>
              </View>
            </FadeUp>
          );
        }
        return (
          <View style={{ marginVertical: 8 }}>
            {Array.isArray(item.sources) && item.sources.length > 0 && (
              <View style={{ flexDirection: 'row', flexWrap: 'wrap', gap: 6, marginBottom: 8 }}>
                {item.sources.slice(0, 4).map((s, i) => {
                  const title = String(s?.title || s?.url || '');
                  const url = typeof s?.url === 'string' ? s.url : null;
                  return (
                    <Pressable
                      key={i}
                      onPress={url ? () => Linking.openURL(url) : undefined}
                      style={({ pressed }) => ({
                        flexDirection: 'row',
                        alignItems: 'center',
                        gap: 6,
                        paddingLeft: 10,
                        paddingRight: 8,
                        paddingVertical: 5,
                        borderRadius: RADIUS_PILL,
                        borderWidth: 1,
                        borderColor: c.borderSoft,
                        backgroundColor: pressed ? c.bgSecondary : c.bg,
                      })}
                    >
                      <Icon name="globe" size={12} color={c.inkSoft} />
                      <Text
                        numberOfLines={1}
                        style={{ maxWidth: 150, fontFamily: FONTS.ui, fontSize: 12, color: c.inkSoft }}
                      >
                        {title}
                      </Text>
                    </Pressable>
                  );
                })}
              </View>
            )}
            {item.content ? (
              <Markdown
                style={markdownStyles(c)}
                onLinkPress={(url) => {
                  Linking.openURL(url);
                  return false;
                }}
              >
                {item.content}
              </Markdown>
            ) : isLast && chat.streaming ? (
              <View style={{ flexDirection: 'row', alignItems: 'center', gap: 8 }}>
                <ActivityIndicator size="small" color={c.inkSoft} />
                <Text style={{ fontFamily: FONTS.ui, fontSize: 13, color: c.inkMuted }}>
                  Pensando…
                </Text>
              </View>
            ) : null}
          </View>
        );
      }}
    />
  );
}

// Compositor (.chat-composer / .chat-input): caja crema redondeada con el
// texto arriba y una barra inferior de acciones.
function Composer({ input, onChangeInput, onSend, onPickModel }) {
  const c = useColors();
  const chat = useChat();
  const insets = useSafeAreaInsets();

  const canSend = !!input.trim() || chat.streaming;

  return (
    <View
      style={{
        paddingHorizontal: 14,
        paddingTop: 6,
        // Con edge-to-edge la barra de navegación se superpone al contenido:
        // sin este inset el compositor quedaba medio tapado por ella.
        paddingBottom: Math.max(insets.bottom, 12),
      }}
    >
      {!!chat.error && (
        <View
          style={{
            marginBottom: 8,
            paddingHorizontal: 12,
            paddingVertical: 8,
            borderRadius: 10,
            backgroundColor: c.dangerSoft,
            flexDirection: 'row',
            alignItems: 'center',
          }}
        >
          <Text style={{ flex: 1, fontFamily: FONTS.ui, fontSize: 13, color: c.danger }}>
            {chat.error}
          </Text>
          <Pressable onPress={chat.clearError} hitSlop={8} style={{ padding: 2 }}>
            <Icon name="x" size={15} color={c.danger} />
          </Pressable>
        </View>
      )}

      <View
        style={{
          backgroundColor: c.bgInput,
          borderRadius: RADIUS_BOX,
          paddingHorizontal: 15,
          paddingTop: 11,
          paddingBottom: 9,
        }}
      >
        <TextInput
          value={input}
          onChangeText={onChangeInput}
          placeholder="Escribe un mensaje…"
          placeholderTextColor={c.inkSoft}
          multiline
          style={{
            maxHeight: 130,
            paddingHorizontal: 6,
            paddingTop: 2,
            paddingBottom: 6,
            fontFamily: FONTS.ui,
            fontSize: 15,
            lineHeight: 21,
            color: c.ink,
          }}
        />
        <View
          style={{
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
            marginTop: 4,
          }}
        >
          <View style={{ flexDirection: 'row', alignItems: 'center', gap: 4, flex: 1 }}>
            {/* Búsqueda web: activo = círculo de tinta (is-active de la web) */}
            <Pressable
              onPress={() => chat.setWebSearch(!chat.webSearch)}
              hitSlop={6}
              style={({ pressed }) => ({
                width: 34,
                height: 34,
                borderRadius: 17,
                alignItems: 'center',
                justifyContent: 'center',
                borderWidth: chat.webSearch ? 0 : 1,
                borderColor: c.borderSoft,
                backgroundColor: chat.webSearch ? c.primary : pressed ? c.pressed : 'transparent',
              })}
            >
              <Icon name="globe" size={18} color={chat.webSearch ? c.onPrimary : c.ink} />
            </Pressable>

            {/* Selector de modelo (.chat-input__model: pill con borde suave) */}
            <Pressable
              onPress={onPickModel}
              style={({ pressed }) => ({
                flexDirection: 'row',
                alignItems: 'center',
                gap: 4,
                marginLeft: 6,
                paddingLeft: 13,
                paddingRight: 9,
                paddingVertical: 6,
                borderRadius: RADIUS_PILL,
                borderWidth: 1,
                borderColor: c.borderSoft,
                backgroundColor: pressed ? c.pressed : 'transparent',
                maxWidth: 190,
              })}
            >
              <Text
                numberOfLines={1}
                style={{ flexShrink: 1, fontFamily: FONTS.ui, fontSize: 13, color: c.ink }}
              >
                {chat.model || 'Sin modelos'}
              </Text>
              <Icon name="chevron-down" size={13} color={c.inkSoft} />
            </Pressable>
          </View>

          {/* Enviar / detener (.chat-input__send: círculo tinta 40px) */}
          <Pressable
            onPress={chat.streaming ? chat.stop : onSend}
            disabled={!canSend}
            style={({ pressed }) => ({
              width: 40,
              height: 40,
              borderRadius: 20,
              backgroundColor: c.primary,
              alignItems: 'center',
              justifyContent: 'center',
              opacity: !canSend ? 0.45 : pressed ? 0.85 : 1,
            })}
          >
            <Icon name={chat.streaming ? 'stop' : 'arrow-up'} size={17} color={c.onPrimary} />
          </Pressable>
        </View>
      </View>
    </View>
  );
}

// Estilos del markdown con los tokens de marca. Un único estilo de código
// (fondo bgSecondary) cubre inline y bloques para que el texto tinta sea
// legible en los dos temas.
function markdownStyles(c) {
  return {
    body: { fontFamily: FONTS.ui, fontSize: 15, lineHeight: 23, color: c.ink },
    heading1: { fontFamily: FONTS.uiSemiBold, fontSize: 20, color: c.ink, marginTop: 8, marginBottom: 4 },
    heading2: { fontFamily: FONTS.uiSemiBold, fontSize: 18, color: c.ink, marginTop: 8, marginBottom: 4 },
    heading3: { fontFamily: FONTS.uiSemiBold, fontSize: 16, color: c.ink, marginTop: 6, marginBottom: 3 },
    strong: { fontFamily: FONTS.uiBold, color: c.ink },
    em: { fontStyle: 'italic', color: c.ink },
    link: { color: c.accentDeep, textDecorationLine: 'underline' },
    bullet_list_icon: { color: c.inkSoft },
    ordered_list_icon: { color: c.inkSoft, fontFamily: FONTS.ui },
    blockquote: {
      backgroundColor: c.bgSecondary,
      borderLeftWidth: 1,
      borderLeftColor: c.borderSoft,
      borderRadius: 6,
      paddingHorizontal: 12,
      marginVertical: 4,
    },
    code_inline: {
      fontFamily: 'monospace',
      fontSize: 13,
      color: c.ink,
      backgroundColor: c.bgSecondary,
      borderRadius: 4,
      paddingHorizontal: 4,
    },
    code_block: {
      fontFamily: 'monospace',
      fontSize: 13,
      color: c.ink,
      backgroundColor: c.bgSecondary,
      borderColor: c.borderSoft,
      borderWidth: 1,
      borderRadius: 12,
      padding: 12,
      marginVertical: 6,
    },
    fence: {
      fontFamily: 'monospace',
      fontSize: 13,
      color: c.ink,
      backgroundColor: c.bgSecondary,
      borderColor: c.borderSoft,
      borderWidth: 1,
      borderRadius: 12,
      padding: 12,
      marginVertical: 6,
    },
    hr: { backgroundColor: c.borderSoft, height: 1, marginVertical: 10 },
    table: { borderColor: c.borderSoft, borderWidth: 1, borderRadius: 10 },
    th: { fontFamily: FONTS.uiSemiBold, fontSize: 13, color: c.ink, padding: 6 },
    td: { fontFamily: FONTS.ui, fontSize: 13, color: c.ink, padding: 6 },
    tr: { borderColor: c.borderSoft },
  };
}
